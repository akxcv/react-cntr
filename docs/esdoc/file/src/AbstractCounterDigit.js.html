<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <base data-ice="baseUrl" href="../../">
  <title data-ice="title">src/AbstractCounterDigit.js | API Document</title>
  <link type="text/css" rel="stylesheet" href="css/style.css">
  <link type="text/css" rel="stylesheet" href="css/prettify-tomorrow.css">
  <script src="script/prettify/prettify.js"></script>
  
  
  <script src="script/manual.js"></script>
</head>
<body class="layout-container" data-ice="rootContainer">

<header>
  <a href="./">Home</a>
  
  <a href="identifiers.html">Reference</a>
  <a href="source.html">Source</a>
  
  <a data-ice="repoURL" href="https://github.com/umbrellio/rollex" class="repo-url-github">Repository</a>
  <div class="search-box">
  <span>
    <img src="./image/search.png">
    <span class="search-input-edge"></span><input class="search-input"><span class="search-input-edge"></span>
  </span>
    <ul class="search-result"></ul>
  </div>
</header>

<nav class="navigation" data-ice="nav"><div>
  <ul>
    
  <li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/AbstractCounterDigit.js~AbstractCounterDigit.html">AbstractCounterDigit</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/AnimatedCounterDigit.js~AnimatedCounterDigit.html">AnimatedCounterDigit</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/CounterSegment.js~CounterSegment.html">CounterSegment</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/StaticCounterDigit.js~StaticCounterDigit.html">StaticCounterDigit</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/index.js~Counter.html">Counter</a></span></span></li>
</ul>
</div>
</nav>

<div class="content" data-ice="content"><h1 data-ice="title">src/AbstractCounterDigit.js</h1>
<pre class="source-code line-number raw-source-code"><code class="prettyprint linenums" data-ice="content">import React from &apos;react&apos;
const { objectOf, any, number, string, func } = React.PropTypes

/**
 * common logic for {@link StaticCounterDigit} and {@link AnimatedCounterDigit}
 */
class AbstractCounterDigit extends React.Component {
  /**
   * propTypes
   * @property {string} digit - digit to display
   * @property {number} height - digit height in pixels
   * @property {number} radix
   * @property {Object} digitMap - a map for transforming particular digits
   * @property {function(digit: number)} digitWrapper - a function for wrapping mapped digits
   */
  static propTypes = {
    digit: string.isRequired,
    height: number.isRequired,
    radix: number.isRequired,
    digitMap: objectOf(any).isRequired,
    digitWrapper: func.isRequired
  }

  /**
   * decorates given digit according to radix, digit map and digit wrapper
   * @param {number} digit
   * @return {string} decoratedDigit
   */
  decorateDigit (digit) {
    const digitString = (digit).toString(this.props.radix)
    return this.props.digitWrapper(this.props.digitMap[digitString] || digitString)
  }
}

export default AbstractCounterDigit
</code></pre>

</div>

<footer class="footer">
  Generated by <a href="https://esdoc.org">ESDoc<span data-ice="esdocVersion">(0.5.2)</span><img src="./image/esdoc-logo-mini-black.png"></a>
</footer>

<script src="script/search_index.js"></script>
<script src="script/search.js"></script>
<script src="script/pretty-print.js"></script>
<script src="script/inherited-summary.js"></script>
<script src="script/test-summary.js"></script>
<script src="script/inner-link.js"></script>
<script src="script/patch-for-local.js"></script>
</body>
</html>
