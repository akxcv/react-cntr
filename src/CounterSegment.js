import React from 'react'
import ReactDOMServer from 'react-dom/server'
import AnimatedCounterDigit from './AnimatedCounterDigit'
import StaticCounterDigit from './StaticCounterDigit'
const { arrayOf, objectOf, number, string, any, func } = React.PropTypes

/**
 * @type {Object}
 * Maximum decimal values for available periods' numbers.
 */
const PERIOD_LIMITS = {
  'seconds': 59,
  'minutes': 59,
  'hours': 23,
  'days': 0
}

/**
 * Counter segment component.
 * @example
 * <CounterSegment
 *   digits={['0', '0']}
 *   period='days'
 *   radix={10}
 *   direction='down'
 *   easingDuration={300}
 *   digitMap={{ '0' => 'o' }}
 *   digitWrapper={(digit) => digit}
 * />
 */
class CounterSegment extends React.Component {
  /**
   * @property {string[]} digits - digits to display
   * @property {string} period
   * @property {number} radix
   * @property {string} direction - counting direction
   * @property {string} easingFunction - easing function to use for digit transitions
   * @property {number} easingDuration - duration of digit transitions
   * @property {Object} digitMap - a map for transforming particular digits
   * @property {function(digit: number)} digitWrapper - a function for wrapping mapped digits
   */
  static propTypes = {
    digits: arrayOf(string).isRequired,
    period: string.isRequired,
    radix: number.isRequired,
    direction: string.isRequired,
    easingFunction: string,
    easingDuration: number.isRequired,
    digitMap: objectOf(any).isRequired,
    digitWrapper: func.isRequired
  }

  /**
   * constructor
   * @param {Object} props
   */
  constructor (props) {
    super(props)

    /**
     * Calculates height of a single digit.
     */
    const testDigit = document.createElement('div')
    testDigit.innerHTML = ReactDOMServer.renderToString(this.props.digitWrapper('0'))
    document.getElementsByTagName('body')[0].appendChild(testDigit)

    /**
     * @type {Object}
     * @property {number} digitHeight - height of a single digit
     */
    this.state = {
      digitHeight: testDigit.clientHeight
    }
    testDigit.remove()
  }

  /**
   * Gets maximum value for period's digit with account for radix.
   * Used for building digit lanes in {@link AnimatedCounterDigit}.
   * @param {number} index - digit's index in number
   * @return {number} maxValue
   */
  getMaxValue (index) {
    const maxValue = PERIOD_LIMITS[this.props.period]
    if (!maxValue) return this.props.radix - 1

    const maxValueString = maxValue.toString(this.props.radix)
    const maxDigitPos = maxValueString.length
    return (index === this.props.digits.length - maxDigitPos)
      ? parseInt(maxValueString[0])
      : this.props.radix - 1
  }

  /**
   * Maps digits to corresponding components according to easing function.
   * @return {ReactElement[]}
   */
  buildDigits () {
    return this.props.digits.map((digit, index) => {
      if (this.props.easingFunction) {
        return (
          <AnimatedCounterDigit
            key={index}
            digit={digit}
            maxValue={this.getMaxValue(index)}
            radix={this.props.radix}
            direction={this.props.direction}
            easingFunction={this.props.easingFunction}
            easingDuration={this.props.easingDuration}
            height={this.state.digitHeight}
            digitMap={this.props.digitMap}
            digitWrapper={this.props.digitWrapper}
          />
        )
      } else {
        return (
          <StaticCounterDigit
            key={index}
            digit={digit}
            radix={this.props.radix}
            height={this.state.digitHeight}
            digitMap={this.props.digitMap}
            digitWrapper={this.props.digitWrapper}
          />
        )
      }
    })
  }

  /**
   * Renders the segment.
   * @return {ReactElement} counter segment
   */
  render () {
    const style = {
      overflow: 'hidden',
      height: `${this.state.digitHeight}px`
    }
    return (
      <div className='rollex-segment'>
        <div className='rollex-digits' style={style}>
          {this.buildDigits()}
        </div>
        <div className='rollex-label'>
          {this.props.period}
        </div>
      </div>
    )
  }
}

export default CounterSegment
