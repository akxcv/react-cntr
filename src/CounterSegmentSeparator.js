import React from 'react'
const { any } = React.PropTypes

/**
 * Separator between {@link CounterSegment} elements.
 * @example
 * <CounterSegmentSeparator content=':' />
 */
class CounterSegmentSeparator extends React.Component {
  /**
   * @property content
   */
  static propTypes = {
    content: any
  }

  /**
   * Renders the separator.
   * @return {ReactElement} separator
   */
  render () {
    if (this.props.content === undefined) return null
    return (
      <div className='rollex-separator'>
        {this.props.content}
      </div>
    )
  }
}

export default CounterSegmentSeparator
