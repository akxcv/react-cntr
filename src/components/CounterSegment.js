import React from 'react'
import PropTypes from 'prop-types'
import AnimatedCounterDigit from './AnimatedCounterDigit'
import StaticCounterDigit from './StaticCounterDigit'

/**
 * @type {Object}
 * Maximum decimal values for available periods' numbers.
 */
const PERIOD_LIMITS = {
  'seconds': 59,
  'minutes': 59,
  'hours': 23,
  'days': 0
}

/**
 * @property {string[]} digits - digits to display
 * @property {string} period
 * @property {number} radix
 * @property {string} direction - counting direction
 * @property {string} easingFunction - easing function to use for digit transitions
 * @property {number} easingDuration - duration of digit transitions
 * @property {Object} digitMap - a map for transforming particular digits
 * @property {function(digit: number)} digitWrapper - a function for wrapping mapped digits
 * @property {string} label
 */
CounterSegment.propTypes = {
  digits: PropTypes.arrayOf(PropTypes.string).isRequired,
  period: PropTypes.string.isRequired,
  radix: PropTypes.number.isRequired,
  direction: PropTypes.string.isRequired,
  easingFunction: PropTypes.string,
  easingDuration: PropTypes.number.isRequired,
  digitMap: PropTypes.object.isRequired,
  digitWrapper: PropTypes.func.isRequired,
  label: PropTypes.string
}

/**
 * Counter segment component.
 * @example
 * <CounterSegment
 *   digits={['0', '0']}
 *   period='days'
 *   radix={10}
 *   direction='down'
 *   easingDuration={300}
 *   digitMap={{ '0' => 'o' }}
 *   digitWrapper={(digit) => digit}
 * />
 */
export default function CounterSegment (props) {
  /**
   * Gets maximum value for period's digit with account for radix.
   * Used for building digit lanes in {@link AnimatedCounterDigit}.
   * @param {number} index - digit's index in number
   * @return {number} maxValue
   */
  function getMaxValue (index) {
    const maxValue = PERIOD_LIMITS[props.period]
    if (!maxValue) return props.radix - 1

    const maxValueString = maxValue.toString(props.radix)
    const maxDigitPos = maxValueString.length
    return (index === props.digits.length - maxDigitPos)
      ? parseInt(maxValueString[0])
      : props.radix - 1
  }

  /**
   * Maps digits to corresponding components according to easing function.
   * @return {ReactElement[]}
   */
  function buildDigits () {
    return props.digits.map((digit, index) => {
      if (props.easingFunction) {
        return (
          <AnimatedCounterDigit
            key={index}
            digit={digit}
            maxValue={getMaxValue(index)}
            radix={props.radix}
            direction={props.direction}
            easingFunction={props.easingFunction}
            easingDuration={props.easingDuration}
            digitMap={props.digitMap}
            digitWrapper={props.digitWrapper}
          />
        )
      } else {
        return (
          <StaticCounterDigit
            key={index}
            digit={digit}
            radix={props.radix}
            digitMap={props.digitMap}
            digitWrapper={props.digitWrapper}
          />
        )
      }
    })
  }

  return (
    <div className='rollex-segment'>
      <div className='rollex-digits' style={{ overflow: 'hidden' }}>
        {buildDigits()}
      </div>
      <div className='rollex-label'>
        {props.label}
      </div>
    </div>
  )
}
