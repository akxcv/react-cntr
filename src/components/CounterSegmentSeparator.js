import React from 'react'
import PropTypes from 'prop-types'

/**
 * Separator between {@link CounterSegment} elements.
 * @example
 * <CounterSegmentSeparator content=':' />
 */
class CounterSegmentSeparator extends React.Component {
  /**
   * @property content
   */
  static propTypes = {
    content: PropTypes.any
  }

  /**
   * Renders the separator.
   * @return {ReactElement} separator
   */
  render () {
    if (this.props.content === undefined) return null
    return (
      <div className='rollex-separator'>
        {this.props.content}
      </div>
    )
  }
}

export default CounterSegmentSeparator
