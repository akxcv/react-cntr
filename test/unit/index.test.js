import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { Counter } from '../../src/'
import CounterSegment from '../../src/CounterSegment'
import CounterSegmentSeparator from '../../src/CounterSegmentSeparator'

describe('initialization', function () {
  it('throws an error when options are provided incorrectly', function () {
    const originalError = console.error
    console.error = () => {} // disable React PropTypes warnings

    expect(
      () => shallow(<Counter />)
    ).toThrowError('provide either "seconds" or "to"')
    expect(
      () => shallow(<Counter from={0} to={1} seconds={2} />)
    ).toThrowError('cannot use "to" and "from" with "seconds"')
    expect(
      () => shallow(<Counter from={0} />)
    ).toThrowError('provide either "seconds" or "to"')
    expect(
      () => shallow(<Counter from={0} seconds={2} />)
    ).toThrowError('cannot use "to" and "from" with "seconds"')
    expect(
      () => shallow(<Counter seconds={2} syncTime />)
    ).toThrowError('"syncTime" must be used with "to"')
    expect(
      () => shallow(<Counter to={0} seconds={2} />)
    ).toThrowError('cannot use "to" and "from" with "seconds"')
    expect(
      () => shallow(<Counter from={1} to={0} />)
    ).toThrowError('"to" must be bigger than "from"')
    expect(
      () => shallow(<Counter seconds={-1} />)
    ).toThrowError('"seconds" must be greater than or equal to zero')
    expect(
      () => shallow(<Counter seconds={0} minDigits={0} />)
    ).toThrowError('"minDigits" must be positive')
    expect(
      () => shallow(<Counter seconds={0} minPeriod='sobaka' />)
    ).toThrowError('"minPeriod" must be one of: day, hour, minute, second')
    expect(
      () => shallow(<Counter seconds={0} maxPeriod='sobaka' />)
    ).toThrowError('"maxPeriod" must be one of: day, hour, minute, second')
    expect(
      () => shallow(<Counter seconds={0} radix={37} />)
    ).toThrowError('"radix" must be between 2 and 36')
    expect(
      () => shallow(<Counter seconds={0} digitWrapper={48} />)
    ).toThrowError('"digitWrapper" must be a function')
    expect(
      () => shallow(<Counter seconds={0} digitMap='111' />)
    ).toThrowError('"digitMap" must be an object')
    expect(
      () => shallow(<Counter to={0} direction='sobaka' />)
    ).toThrowError('"direction" must be either up or down')

    console.error = originalError
  })

  it('initializes when options are correct', function () {
    expect(
      () => shallow(<Counter to={0} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter from={0} to={1} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} minDigits={1} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} maxDigits={3} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} minPeriod='minute' />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} maxPeriod='hour' />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter from={0} to={1} syncTime />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={0} radix={12} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={0} digitWrapper={() => null} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={0} digitMap={{}} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter to={9} direction='up' />)
    ).not.toThrow()
  })
})

describe('rendering', function () {
  const to =
    (1000 * 60 * 60 * 24 * 200) +
    (1000 * 60 * 60 * 6) +
    (1000 * 60 * 35) +
    (1000 * 54)

  it('matches snapshot', function () {
    const component = shallow(<Counter from={0} to={1} />)
    const tree = shallowToJson(component)
    expect(tree).toMatchSnapshot()
  })

  it('renders a CounterSegment for each segment', function () {
    const component = shallow(<Counter from={0} to={1} />)
    expect(component.find(CounterSegment).length).toEqual(4)
  })

  it('passes digits correctly', function () {
    const component = shallow(<Counter from={0} to={to} />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['2', '0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['5', '4'])
  })

  it('works with "up" direction', function () {
    const component = shallow(<Counter from={0} to={to} direction='up' />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '0'])
    expect(counterSegments.at(2).props().digits).toEqual(['0', '0'])
    expect(counterSegments.at(3).props().digits).toEqual(['0', '0'])
  })

  test('minDigits', function () {
    var component = shallow(<Counter from={0} to={to} minDigits={1} />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['2', '0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['6'])
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['5', '4'])

    component = shallow(<Counter from={0} to={to} minDigits={3} />)
    counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['2', '0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['0', '3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['0', '5', '4'])
  })

  test('maxDigits', function () {
    var component = shallow(<Counter from={0} to={to} maxDigits={1} />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['9'])
    expect(counterSegments.at(1).props().digits).toEqual(['6'])
    expect(counterSegments.at(2).props().digits).toEqual(['9'])
    expect(counterSegments.at(3).props().digits).toEqual(['9'])

    component = shallow(<Counter from={0} to={to} maxDigits={2} />)
    counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['9', '9'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['5', '4'])
  })

  test('minPeriod', function () {
    const component = shallow(<Counter from={0} to={to} minPeriod='minute' />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(3)
    expect(counterSegments.at(2).props().period).toBe('minutes')
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
  })

  test('maxPeriod', function () {
    var component = shallow(<Counter from={0} to={to} maxPeriod='hour' />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(3)
    expect(counterSegments.at(0).props().period).toBe('hours')
    expect(counterSegments.at(0).props().digits).toEqual(['4', '8', '0', '6'])

    var component = shallow(<Counter from={0} to={to} maxPeriod='minute' />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(2)
    expect(counterSegments.at(0).props().period).toBe('minutes')
    expect(counterSegments.at(0).props().digits).toEqual(['2', '8', '8', '3', '9', '5'])

    var component = shallow(<Counter from={0} to={to} maxPeriod='second' />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(1)
    expect(counterSegments.at(0).props().period).toBe('seconds')
    expect(counterSegments.at(0).props().digits).toEqual(['1', '7', '3', '0', '3', '7', '5', '4'])
  })

  test('radix', function () {
    var component = shallow(<Counter from={0} to={to} radix={12} />)
    var counterSegments = component.find(CounterSegment)

    expect(counterSegments.at(0).props().digits).toEqual(['1', '4', '8'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['2', 'b'])
    expect(counterSegments.at(3).props().digits).toEqual(['4', '6'])
  })

  it('passes digitWrapper to segments', function () {
    const digitWrapper = (digit) => digit + '0'
    const component = shallow(<Counter from={0} to={to} digitWrapper={digitWrapper} />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digitWrapper).toEqual(digitWrapper)
    expect(counterSegments.at(1).props().digitWrapper).toEqual(digitWrapper)
    expect(counterSegments.at(2).props().digitWrapper).toEqual(digitWrapper)
    expect(counterSegments.at(3).props().digitWrapper).toEqual(digitWrapper)
  })

  it('passes digitMap to segments', function () {
    const digitMap = { '0': 'o' }
    const component = shallow(<Counter from={0} to={to} digitMap={digitMap} />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digitMap).toEqual(digitMap)
    expect(counterSegments.at(1).props().digitMap).toEqual(digitMap)
    expect(counterSegments.at(2).props().digitMap).toEqual(digitMap)
    expect(counterSegments.at(3).props().digitMap).toEqual(digitMap)
  })

  describe('segment labels', function () {
    it('passes default labels', function () {
      const component = shallow(<Counter to={1} />)
      const counterSegments = component.find(CounterSegment)
      expect(counterSegments.at(0).props().label).toEqual('days')
      expect(counterSegments.at(1).props().label).toEqual('hours')
      expect(counterSegments.at(2).props().label).toEqual('minutes')
      expect(counterSegments.at(3).props().label).toEqual('seconds')
    })

    it('passes labels when given a map', function () {
      const labelMap = {
        days: 'DD',
        hours: 'HH',
        minutes: 'MM',
        seconds: 'SS'
      }
      const component = shallow(<Counter to={1} labels={labelMap} />)
      const segments = component.find(CounterSegment)
      expect(segments.at(0).props().label).toEqual('DD')
      expect(segments.at(1).props().label).toEqual('HH')
      expect(segments.at(2).props().label).toEqual('MM')
      expect(segments.at(3).props().label).toEqual('SS')
    })

    it('passes labels when given a function', function () {
      const labelFunc = function (period, number) {
        if (number % 10 === 1) {
          return period.slice(0, -1)
        } else {
          return period
        }
      }
      const component = shallow(<Counter seconds={61} labels={labelFunc} />)
      const segments = component.find(CounterSegment)
      expect(segments.at(0).props().label).toEqual('days')
      expect(segments.at(1).props().label).toEqual('hours')
      expect(segments.at(2).props().label).toEqual('minute')
      expect(segments.at(3).props().label).toEqual('second')
    })
  })

  test('separator', function () {
    const component = shallow(<Counter seconds={10} separator=':' />)
    const separators = component.find(CounterSegmentSeparator)
    expect(separators.at(0).props().content).toEqual(':')
    expect(separators.at(1).props().content).toEqual(':')
    expect(separators.at(2).props().content).toEqual(':')
    expect(separators.at(3).props().content).toEqual(undefined)
  })
})

describe('state and props', function () {
  test('default state', function () {
    jest.spyOn(Date.prototype, 'getTime').mockReturnValue(1234)
    const to = 1234 +
      (1000 * 60 * 60 * 24 * 2) +
      (1000 * 60 * 60 * 6) +
      (1000 * 60 * 35) +
      (1000 * 54)
    const component = shallow(<Counter to={to} />)
    expect(component.state()).toEqual({
      timeDiff: to - 1234,
      minDigits: 2,
      startTime: 1234,
      from: 1234,
      initialTimeDiff: to - 1234,
      periods: ['days', 'hours', 'minutes', 'seconds'],
      cssClasses: 'rollex rollex-static',
      numbers: {
        days: 2,
        hours: 6,
        minutes: 35,
        seconds: 54
      }
    })
  })

  test('default props', function () {
    const component = mount(<Counter from={0} to={1} />)
    expect(component.props()).toEqual({
      from: 0,
      to: 1,
      interval: 1000,
      minPeriod: 'second',
      maxPeriod: 'day',
      radix: 10,
      direction: 'down',
      frozen: false,
      easingFunction: null,
      easingDuration: 300,
      digitMap: {},
      digitWrapper: expect.any(Function)
    })
  })

  it('allows to set props', function () {
    const digitWrapper = (digit) => <div>{digit}</div>
    const component = mount(
      <Counter from={10} to={20} interval={897}
        minDigits={3} maxDigits={4}
        minPeriod='minute' maxPeriod='hour'
        syncTime radix={8} direction='up'
        easingFunction='myEasingFn' easingDuration={123}
        digitMap={{ '0': 'o' }} digitWrapper={digitWrapper}
        separator=':'
      />
    )
    expect(component.props()).toEqual({
      from: 10,
      to: 20,
      interval: 897,
      minDigits: 3,
      maxDigits: 4,
      minPeriod: 'minute',
      maxPeriod: 'hour',
      syncTime: true,
      easingFunction: 'myEasingFn',
      easingDuration: 123,
      radix: 8,
      frozen: false,
      direction: 'up',
      digitMap: { '0': 'o' },
      digitWrapper: digitWrapper,
      separator: ':'
    })
  })
})

describe('counting', function () {
  beforeEach(function () {
    jest.clearAllTimers()
    jest.useFakeTimers()
    delete (window.rollexIntervals)
    // FIXME: the following line is required. It shouldn't be. Something's wrong with global state.
    mount(<Counter seconds={10000} />)
  })

  it('starts to count when mounted', function () {
    const component = mount(<Counter from={0} to={10000} />)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
    expect(component.state().numbers.seconds).toBe(5)
  })

  it('works with "seconds" prop', function () {
    const component = mount(<Counter seconds={10} />)
    expect(component.state().numbers.seconds).toBe(10)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
    expect(component.state().numbers.seconds).toBe(5)
  })

  it('works without "from" prop', function () {
    jest.spyOn(Date.prototype, 'getTime').mockReturnValue(78781234)
    const to = new Date().getTime() + 10000
    const component = mount(<Counter to={to} />)
    expect(component.state().timeDiff).toBe(10000)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
  })

  it('does not count when "frozen" is true', function () {
    const component = mount(<Counter seconds={10} frozen />)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(10000)
    expect(component.state().numbers.seconds).toBe(10)
  })

  it('stops to count when stopped', function () {
    const component = mount(<Counter from={0} to={10000} />)
    jest.runTimersToTime(5000)
    expect(component.state().numbers.seconds).toBe(5)

    component.instance().stop()

    jest.runTimersToTime(10000)
    expect(component.state().numbers.seconds).toBe(5)
  })

  it('stops to count when reached zero', function () {
    const component = mount(<Counter from={0} to={10000} />)
    jest.runTimersToTime(10000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(0)

    jest.runTimersToTime(10000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(0)
  })

  it('stops to count if frozen prop is updated to true', function () {
    const component = mount(<Counter from={0} to={10000} />)
    component.instance().stop = jest.genMockFunction()
    component.instance().forceUpdate()
    component.setProps({
      frozen: true
    }, () => expect(component.instance().stop.mock.calls.length).toBe(1))

    const frozenComponent = mount(<Counter from={0} to={10000} frozen />)
    frozenComponent.instance().stop = jest.genMockFunction()
    frozenComponent.instance().forceUpdate()
    frozenComponent.setProps({
      frozen: true
    }, () => expect(frozenComponent.instance().stop.mock.calls.length).toBe(0))
  })

  it('starts to count if frozen prop is updated to false', function () {
    const component = mount(<Counter from={0} to={10000} />)
    component.instance().start = jest.genMockFunction()
    component.instance().forceUpdate()
    component.setProps({
      frozen: false
    }, () => expect(component.instance().start.mock.calls.length).toBe(0))

    const frozenComponent = mount(<Counter from={0} to={10000} frozen />)
    frozenComponent.instance().start = jest.genMockFunction()
    frozenComponent.instance().forceUpdate()
    frozenComponent.setProps({
      frozen: false
    }, () => expect(frozenComponent.instance().start.mock.calls.length).toBe(1))
  })

  it('syncs time when syncTime is set', function () {
    const component = mount(<Counter from={0} to={10000} syncTime />)
    const newDate = component.state().startTime + 7000
    jest.spyOn(Date.prototype, 'getTime').mockReturnValue(newDate)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(3000)
  })

  it('works with "up" direction', function () {
    const component = mount(<Counter seconds={10} direction='up' />)
    expect(component.state().timeDiff).toBe(10000)
    expect(component.state().numbers.seconds).toBe(0)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
    expect(component.state().numbers.seconds).toBe(5)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(10)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(10)
  })
})

describe('CSS classes', function () {
  it('applies rollex and rollex-static classes to static counters', function () {
    const component = shallow(<Counter seconds={0} />)
    expect(component.find('.rollex').at(0).hasClass('rollex-static')).toBe(true)
  })

  it('applies rollex and rollex-animated classes to animated counters', function () {
    const component = shallow(<Counter seconds={0} easingFunction='ease-in' />)
    expect(component.find('.rollex').at(0).hasClass('rollex-animated')).toBe(true)
  })

  it('applies rollex-frozen class to frozen counters', function () {
    const staticComponent = shallow(<Counter seconds={0} frozen />)
    expect(staticComponent.find('.rollex').at(0).hasClass('rollex-frozen')).toBe(true)
    const animatedComponent = shallow(<Counter seconds={0} easingFunction='ease-in' frozen />)
    expect(animatedComponent.find('.rollex').at(0).hasClass('rollex-frozen')).toBe(true)
  })
})
