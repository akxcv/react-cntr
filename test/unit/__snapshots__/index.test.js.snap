// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering matches snapshot 1`] = `
<div
  className="rollex rollex-static"
>
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    direction="down"
    easingDuration={300}
    easingFunction={null}
    label="days"
    period="days"
    radix={10}
  />
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    direction="down"
    easingDuration={300}
    easingFunction={null}
    label="hours"
    period="hours"
    radix={10}
  />
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    direction="down"
    easingDuration={300}
    easingFunction={null}
    label="minutes"
    period="minutes"
    radix={10}
  />
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    direction="down"
    easingDuration={300}
    easingFunction={null}
    label="seconds"
    period="seconds"
    radix={10}
  />
</div>
`;
